File: Palace/Book/Models/TPPBookRegistryRecord.m
#import "TPPBook.h"
#import "TPPBookLocation.h"
#import "TPPBookRegistryRecord.h"
#import "TPPNull.h"
#import "TPPOPDS.h"
#import "Palace-Swift.h"

@interface TPPBookRegistryRecord ()

@property (nonatomic) TPPBook *book;
@property (nonatomic) TPPBookLocation *location;
@property (nonatomic) TPPBookState state;
@property (nonatomic) NSString *fulfillmentId;
@property (nonatomic) NSArray<TPPReadiumBookmark *> *readiumBookmarks;
@property (nonatomic) NSArray<TPPBookLocation *> *genericBookmarks;

@end

static NSString *const BookKey = @"metadata";
static NSString *const StateKey = @"state";
static NSString *const FulfillmentIdKey = @"fulfillmentId";
static NSString *const ReadiumBookmarksKey = @"bookmarks";
static NSString *const GenericBookmarksKey = @"genericBookmarks";

@implementation TPPBookRegistryRecord

- (instancetype)initWithBook:(TPPBook *const)book
                    location:(TPPBookLocation *const)location
                       state:(TPPBookState)state
               fulfillmentId:(NSString *)fulfillmentId
            readiumBookmarks:(NSArray<TPPReadiumBookmark *> *)readiumBookmarks
            genericBookmarks:(NSArray<TPPBookLocation *> *)genericBookmarks
{
  self = [super init];
  if(!self) return nil;
  
  if(!book) {
    @throw NSInvalidArgumentException;
  }
  
  self.book = book;
  self.location = location;
  self.state = state;
  self.fulfillmentId = fulfillmentId;
  self.readiumBookmarks = readiumBookmarks;
  self.genericBookmarks = genericBookmarks;

  if (!book.defaultAcquisition) {
    // Since the book has no default acqusition, there is no reliable way to
    // determine if the book is on hold (although it may be), nor is there any
    // way to download the book if it is available. As such, we give the book a
    // special "unsupported" state which will allow other parts of the app to
    // ignore it as appropriate. Unsupported books should generally only appear
    // when a user has checked out or reserved a book in an unsupported format
    // using another app.
    self.state = TPPBookStateUnsupported;
    return self;
  }

  // FIXME: The logic below is confusing at best. Upon initial inspection, it's
  // unclear why `book.state` needs to be "fixed" in this initializer. If said
  // fixing is appropriate, a rationale should be added here.

  // If the book availability indicates that the book is held, make sure the state
  // reflects that.
  __block BOOL actuallyOnHold = NO;
  [book.defaultAcquisition.availability
   matchUnavailable:nil
   limited:nil
   unlimited:nil
   reserved:^(__unused TPPOPDSAcquisitionAvailabilityReserved *_Nonnull reserved) {
     self.state = TPPBookStateHolding;
     actuallyOnHold = YES;
   } ready:^(__unused TPPOPDSAcquisitionAvailabilityReady *_Nonnull ready) {
     self.state = TPPBookStateHolding;
     actuallyOnHold = YES;
   }];

  if (!actuallyOnHold) {
    // Set the correct non-holding state.
    if (self.state == TPPBookStateHolding || self.state == TPPBookStateUnsupported)
    {
      // Since we're not in some download-related state and we're not unregistered,
      // we must need to be downloaded.
      self.state = TPPBookStateDownloadNeeded;
    }
  }
  
  return self;
}

- (instancetype)initWithDictionary:(NSDictionary *)dictionary
{
  self = [super init];
  if(!self) return nil;
  
  self.book = [[TPPBook alloc] initWithDictionary:dictionary[BookKey]];
  if(![self.book isKindOfClass:[TPPBook class]]) return nil;
  
  self.location = [[TPPBookLocation alloc]
                   initWithDictionary:TPPNullToNil(dictionary[TPPBookmarkDictionaryRepresentation.locationKey])];
  if(self.location && ![self.location isKindOfClass:[TPPBookLocation class]]) return nil;
  
  NSNumber *state = [TPPBookStateHelper bookStateFromString:dictionary[StateKey]];
  if (state) {
    self.state = state.integerValue;
  } else {
    [TPPErrorLogger logErrorWithCode:TPPErrorCodeUnknownBookState
                              summary:@"Invalid nil state during BookRegistryRecord init"
                             metadata:@{
                               @"Input dict": dictionary ?: @"N/A"
                             }];
    @throw NSInvalidArgumentException;
  }
  
  self.fulfillmentId = TPPNullToNil(dictionary[FulfillmentIdKey]);
  
  NSMutableArray<TPPReadiumBookmark *> *readiumBookmarks = [NSMutableArray array];
  for (NSDictionary *dict in TPPNullToNil(dictionary[ReadiumBookmarksKey])) {
    [readiumBookmarks addObject:[[TPPReadiumBookmark alloc] initWithDictionary:dict]];
  }
  self.readiumBookmarks = readiumBookmarks;

  NSMutableArray<TPPBookLocation *> *genericBookmarks = [NSMutableArray array];
  for (NSDictionary *dict in TPPNullToNil(dictionary[GenericBookmarksKey])) {
    [genericBookmarks addObject:[[TPPBookLocation alloc] initWithDictionary:dict]];
  }
  self.genericBookmarks = genericBookmarks;
  
  return self;
}

- (NSDictionary *)dictionaryRepresentation
{
  NSMutableArray *readiumBookmarks = [NSMutableArray array];
  for (TPPReadiumBookmark *readiumBookmark in self.readiumBookmarks) {
    [readiumBookmarks addObject:readiumBookmark.dictionaryRepresentation];
  }

  NSMutableArray *genericBookmarks = [NSMutableArray array];
  for (TPPBookLocation *genericBookmark in self.genericBookmarks) {
    [genericBookmarks addObject:genericBookmark.dictionaryRepresentation];
  }
  
  return @{BookKey: [self.book dictionaryRepresentation],
           TPPBookmarkDictionaryRepresentation.locationKey: TPPNullFromNil([self.location dictionaryRepresentation]),
           StateKey: [TPPBookStateHelper stringValueFromBookState:self.state],
           FulfillmentIdKey: TPPNullFromNil(self.fulfillmentId),
           ReadiumBookmarksKey: TPPNullFromNil(readiumBookmarks),
           GenericBookmarksKey: TPPNullFromNil(genericBookmarks)};
}

- (instancetype)recordWithBook:(TPPBook *const)book
{
  return [[[self class] alloc] initWithBook:book location:self.location state:self.state fulfillmentId:self.fulfillmentId readiumBookmarks:self.readiumBookmarks genericBookmarks:self.genericBookmarks];
}

- (instancetype)recordWithLocation:(TPPBookLocation *const)location
{
  return [[[self class] alloc] initWithBook:self.book location:location state:self.state fulfillmentId:self.fulfillmentId readiumBookmarks:self.readiumBookmarks genericBookmarks:self.genericBookmarks];
}

- (instancetype)recordWithState:(TPPBookState const)state
{
  return [[[self class] alloc] initWithBook:self.book location:self.location state:state fulfillmentId:self.fulfillmentId readiumBookmarks:self.readiumBookmarks genericBookmarks:self.genericBookmarks];
}

- (instancetype)recordWithFulfillmentId:(NSString *)fulfillmentId
{
  return [[[self class] alloc] initWithBook:self.book location:self.location state:self.state fulfillmentId:fulfillmentId readiumBookmarks:self.readiumBookmarks genericBookmarks:self.genericBookmarks];
}
  
- (instancetype)recordWithReadiumBookmarks:(NSArray<TPPReadiumBookmark *> *)bookmarks
{
  return [[[self class] alloc] initWithBook:self.book location:self.location state:self.state fulfillmentId:self.fulfillmentId readiumBookmarks:bookmarks genericBookmarks:self.genericBookmarks];
}

- (instancetype)recordWithGenericBookmarks:(NSArray<TPPBookLocation *> *)bookmarks
{
  return [[[self class] alloc] initWithBook:self.book location:self.location state:self.state fulfillmentId:self.fulfillmentId readiumBookmarks:self.readiumBookmarks genericBookmarks:bookmarks];
}
  
@end

File: Palace/Book/Models/TPPBookRegistryRecord.swift
//
//  TPPBookRegistryRecord.swift
//  Palace
//
//  Created by Vladimir Fedorov on 09.11.2022.
//  Copyright © 2022 The Palace Project. All rights reserved.
//

import Foundation


/// An element of `TPPBookRegistry`
@objcMembers
class TPPBookRegistryRecord: NSObject {
  var book: TPPBook
  var location: TPPBookLocation?
  var state: TPPBookState
  var fulfillmentId: String?
  var readiumBookmarks: [TPPReadiumBookmark]?
  var genericBookmarks: [TPPBookLocation]?
  
  /// Creates a registry record with the specified state.
  /// 
  /// **Important:** This initializer preserves the passed `state` value. If you need to derive
  /// the initial state from book availability, use `TPPBookRegistryRecord.deriveInitialState(for:)` first.
  ///
  /// - Parameters:
  ///   - book: The book metadata
  ///   - location: Reading position (optional)
  ///   - state: The book state - this value is preserved as-is
  ///   - fulfillmentId: DRM fulfillment identifier (optional)
  ///   - readiumBookmarks: Readium-format bookmarks
  ///   - genericBookmarks: Generic location bookmarks
  init(book: TPPBook, location: TPPBookLocation? = nil, state: TPPBookState, fulfillmentId: String? = nil, readiumBookmarks: [TPPReadiumBookmark]? = [], genericBookmarks: [TPPBookLocation]? = []) {
    self.book = book
    self.location = location
    self.fulfillmentId = fulfillmentId
    self.readiumBookmarks = readiumBookmarks
    self.genericBookmarks = genericBookmarks
    
    // Preserve the passed state - do NOT override based on availability
    // The caller is responsible for determining the correct state
    self.state = state

    super.init()
  }
  
  /// Derives the appropriate initial state for a newly borrowed/discovered book based on its availability.
  /// Use this when adding a book to the registry for the first time without a known state.
  ///
  /// - Parameter book: The book to derive state for
  /// - Returns: The appropriate initial state based on availability
  static func deriveInitialState(for book: TPPBook) -> TPPBookState {
    guard let defaultAcquisition = book.defaultAcquisition else {
      // No acquisition means unsupported format
      return .unsupported
    }
    
    var derivedState: TPPBookState = .downloadNeeded
    
    defaultAcquisition.availability.matchUnavailable { _ in
      derivedState = .downloadNeeded
    } limited: { _ in
      derivedState = .downloadNeeded
    } unlimited: { _ in
      derivedState = .downloadNeeded
    } reserved: { _ in
      derivedState = .holding
    } ready: { _ in
      derivedState = .holding
    }
    
    return derivedState
  }
  
  init?(record: TPPBookRegistryData) {
    guard let bookObject = record.object(for: .book),
          let book = TPPBook(dictionary: bookObject),
          let stateString = record.value(for: .state) as? String,
          let state = TPPBookState(stateString)
            
    else {
      return nil
    }
    self.book = book
    self.state = state
    self.fulfillmentId = record.value(for: .fulfillmentId) as? String
    if let location = record.object(for: .location) {
        self.location = TPPBookLocation(dictionary: location)
    }
    if let recordReadiumBookmarks = record.array(for: .readiumBookmarks) {
      self.readiumBookmarks = recordReadiumBookmarks.compactMap { TPPReadiumBookmark(dictionary: $0 as NSDictionary) }
    }
    if let recordGenericBookmarks = record.array(for: .genericBookmarks) {
      self.genericBookmarks = recordGenericBookmarks.compactMap { TPPBookLocation(dictionary: $0) }
    }
  }
  
  var dictionaryRepresentation: [String: Any] {
    var dictionary = TPPBookRegistryData()
    dictionary.setValue(book.dictionaryRepresentation(), for: .book)
    dictionary.setValue(state.stringValue(), for: .state)
    dictionary.setValue(fulfillmentId, for: .fulfillmentId)
    dictionary.setValue(self.location?.dictionaryRepresentation, for: .location)
    dictionary.setValue(readiumBookmarks?.compactMap { $0.dictionaryRepresentation as? [String: Any] }, for: .readiumBookmarks)
    dictionary.setValue(genericBookmarks?.map { $0.dictionaryRepresentation }, for: .genericBookmarks)
    return dictionary
  }
}
